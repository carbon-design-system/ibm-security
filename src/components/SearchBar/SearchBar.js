/**
 * @file SearchBar.
 * @copyright IBM Security 2019 - 2021
 */

import React from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';
import Button from '../Button';
import Search from '../Search';
import MultiSelect from '../MultiSelect';

import { getComponentNamespace } from '../../globals/namespace';

export const namespace = getComponentNamespace('search-bar');

/**
 * Custom scope validator for props that are only required when scopes has been provided.
 * @param {object} props Object containing passed in props.
 * @param {string} propName Name of prop the validator is for.
 * @param {string} componentName Name of component.
 * @param  {...any} rest Rest of parameters to pass down to string prop validator.
 * @returns {Error|null} Error if validation failed otherwise nothing.
 */
const conditionalScopePropValidator = (
  props,
  propName,
  componentName,
  ...rest
) => {
  if (props.scopes && props.scopes.length > 0 && !props[propName]) {
    return new Error(
      `Required \`${propName}\` when \`scopes\` prop type is supplied to \`${componentName}\`. Validation failed.`
    );
  }

  return PropTypes.string(props, propName, componentName, ...rest);
};

export default class SearchBar extends React.Component {
  static propTypes = {
    /** @type {string} Optional additional class name. */
    className: PropTypes.string,

    /** @type {string} Search query value. */
    value: PropTypes.string,

    /** @type {Function} Function handler for when the user submits a search. */
    onSubmit: PropTypes.func,

    /** @type {Function} Function handler for when the user changes their query search. */
    onChange: PropTypes.func,

    /** @type {string} The label text for the search text input. */
    clearButtonLabelText: PropTypes.string.isRequired,

    /** @type {string} Placeholder text to be displayed in the search input. */
    placeHolderText: PropTypes.string.isRequired,

    /** @type {string} The label text for the search text input. */
    labelText: PropTypes.string.isRequired,

    /** @type {string} The label text for the search submit button. */
    submitLabel: PropTypes.string.isRequired,

    /** @type {Array<any>} Array of allowed search scopes. */
    scopes: PropTypes.arrayOf(
      PropTypes.oneOfType([PropTypes.string, PropTypes.object])
    ),

    /** @type {Array<any> Array of initially selected search scopes. */
    selectedScopes: PropTypes.arrayOf(
      PropTypes.oneOfType([PropTypes.string, PropTypes.object])
    ),

    /** @type {function} Function to get the text for each sscope to display in dropdown. */
    scopeToString: PropTypes.func,

    /** @type {string} The name text for the search scope type. */
    // eslint-disable-next-line react/require-default-props
    scopesTypeLabel: conditionalScopePropValidator,

    /** @type {func} Callback function for translating MultiSelect's child ListBoxMenuIcon SVG title. */
    translateWithId: PropTypes.func, // eslint-disable-line react/require-default-props

    /**
     * Optional custom sorting algorithm for an array of scope items.
     * By default, scope items are sorted in ascending alphabetical order,
     * with "selected" items moved to the start of the scope items array.
     */
    sortItems: PropTypes.func, // eslint-disable-line react/require-default-props

    /**
     * Provide accessible label text for the scopes MultiSelect.
     */
    titleText: PropTypes.string,

    /**
     * Whether or not the scopes MultiSelect label is visually hidden.
     */
    hideScopesLabel: PropTypes.bool,
  };

  static defaultProps = {
    className: '',
    value: '',
    onSubmit: () => {},
    onChange: () => {},
    scopeToString: () => {},
    scopes: [],
    selectedScopes: [],
    titleText: 'Scopes multiselect',
    hideScopesLabel: true,
  };

  /**
   * Handler for search input changes that calls onChange prop.
   * @param {KeyboardEvent} event Event object from input change.
   */
  handleInputChange = event => {
    const { value } = event.target;
    const eventObject = { value };

    if (this.props.scopes.length > 0) {
      eventObject.selectedScopes = this.props.selectedScopes;
    }

    this.props.onChange(eventObject);
  };

  /**
   * Handler for form submit that calls onSubmit prop.
   * @param {Event} event Submit event generated.
   */
  handleSubmit = event => {
    event.preventDefault();
    const eventObject = { value: this.props.value };

    if (this.props.scopes.length > 0) {
      eventObject.selectedScopes = this.props.selectedScopes;
    }

    this.props.onSubmit(eventObject);
  };

  /**
   * Handler for when scope selection changes that calls onChangeProp.
   * @param {{selectedItems: Array<any>}} {selectedItems} Object containing array of selected items.
   */
  handleSearchScopeChange = ({ selectedItems }) => {
    this.props.onChange({
      selectedScopes: selectedItems,
      value: this.props.value,
    });
  };

  /**
   * Renders scope selector dropdown.
   * @returns {ReactElement} Scope selector.
   */
  renderScopeSelector() {
    const {
      scopes,
      scopeToString,
      scopesTypeLabel,
      selectedScopes,
      translateWithId,
      sortItems,
      titleText,
    } = this.props;

    if (scopes.length === 0) return null;

    return (
      <MultiSelect
        id={`${namespace}__multi-select`}
        name="search-scopes"
        className={`${namespace}__scopes`}
        label={scopesTypeLabel}
        onChange={this.handleSearchScopeChange}
        initialSelectedItems={selectedScopes}
        items={scopes}
        itemToString={scopeToString}
        translateWithId={translateWithId}
        sortItems={sortItems}
        titleText={titleText}
      />
    );
  }

  render() {
    const {
      labelText,
      placeHolderText,
      submitLabel,
      value,
      className,
      scopes,
      selectedScopes,
      clearButtonLabelText,
      hideScopesLabel,
    } = this.props;

    return (
      <form
        className={classnames(namespace, className, {
          [`${namespace}--hide-scopes-label`]: hideScopesLabel,
        })}
        onSubmit={this.handleSubmit}
      >
        {this.renderScopeSelector()}
        <Search
          className={`${namespace}__input`}
          closeButtonLabelText={clearButtonLabelText}
          labelText={labelText}
          name="search-input"
          onChange={this.handleInputChange}
          placeholder={placeHolderText}
          value={value}
        />
        <Button
          name="search-submit"
          kind="primary"
          type="submit"
          className={`${namespace}__submit`}
          disabled={
            !value ||
            value.length === 0 ||
            (scopes.length !== 0 && selectedScopes.length === 0)
          }
        >
          {submitLabel}
        </Button>
      </form>
    );
  }
}
