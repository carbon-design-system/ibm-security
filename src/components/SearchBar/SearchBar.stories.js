/**
 * @file Search bar stories.
 * @copyright IBM Security 2019
 */

import { withA11y } from '@storybook/addon-a11y';
import { action } from '@storybook/addon-actions';
import { text } from '@storybook/addon-knobs';
import { storiesOf } from '@storybook/react';

import React from 'react';
import { compose, getDisplayName, withHandlers, withState } from 'recompose';

import { patterns } from '../../../.storybook';

import { SearchBar } from '../..';
import props from './_mocks_';

const SearchBarWithStateHandlers = compose(
  withState(
    'selectedScopes',
    'updateSelectedScopes',
    ({ selectedScopes = [] }) => selectedScopes
  ),
  withState('value', 'updateValue', ({ value = '' }) => value),
  withHandlers({
    onChange: ({
      onChange,
      updateSelectedScopes,
      updateValue,
    }) => searchObject => {
      const { selectedScopes, value } = searchObject;
      onChange(searchObject);

      updateSelectedScopes(selectedScopes);
      updateValue(value);
    },
  })
)(SearchBar);

SearchBarWithStateHandlers.displayName = getDisplayName(SearchBar);
SearchBarWithStateHandlers.__docgenInfo = SearchBar.__docgenInfo;

const {
  allSelectedScopesLabel,
  clearButtonLabelText,
  labelText,
  placeHolderText,
  scopes,
  selectedScopes,
  scopesTypeLabel,
  submitLabel,
  value,
} = props;

const storyProps = {
  regular: () => ({
    submitLabel: text('Submit label (submitLabel)', submitLabel),
    placeHolderText: text('Placeholder (placeholderText)', placeHolderText),
    labelText,
    clearButtonLabelText,
    onChange: action('onChange'),
    onSubmit: action('onSubmit'),
  }),
  scopes: () => ({
    scopes,
    scopesTypeLabel: text(
      'Scopes type label (scopesTypeLabel)',
      scopesTypeLabel
    ),
    allSelectedScopesLabel: text(
      'All selected scopes label (allSelectedScopesLabel)',
      allSelectedScopesLabel
    ),
    scopeToString: ({ name }) => name,
  }),
  translationIds: {
    'close.menu': 'Close the menu',
    'open.menu': 'Open the menu',
    'clear.all': 'Clear all scopes',
    'clear.selection': 'Clear scope',
  },
};

const { regular, scopes: scopesProps, translationIds } = storyProps;

storiesOf(patterns('SearchBar'), module)
  .addDecorator(withA11y)
  .add('Default', () => <SearchBarWithStateHandlers {...regular()} />)
  .add('Initial value', () => (
    <SearchBarWithStateHandlers {...regular()} value={value} />
  ))
  .add('Scopes', () => (
    <SearchBarWithStateHandlers
      {...regular()}
      {...scopesProps()}
      translateWithId={id => translationIds[id]}
    />
  ))
  .add('Selected scopes', () => (
    <SearchBarWithStateHandlers
      {...regular()}
      {...scopesProps()}
      selectedScopes={selectedScopes}
      translateWithId={id => translationIds[id]}
    />
  ));
