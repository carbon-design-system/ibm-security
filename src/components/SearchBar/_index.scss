////
/// Search bar component.
/// @group search-bar
/// @copyright IBM Security 2019
////

@import '@carbon/layout/scss/mini-unit';
@import '@carbon/themes/scss/tokens';

@import 'carbon-components/scss/globals/scss/vars';

@import '../Button/index';
@import '../MultiSelect/index';
@import '../Search/index';

@import '../Component/mixins';

@include security--component($name: search-bar) {
  display: flex;

  &__scopes.#{$prefix}--list-box {
    flex: 0 0 carbon--mini-units(28);
    border: 0;

    &,
    .#{$prefix}--list-box__field {
      height: 100%;
      max-height: unset;
    }
  }
}
