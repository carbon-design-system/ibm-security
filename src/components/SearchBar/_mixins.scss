////
/// Search bar mixins.
/// @group search-bar
/// @copyright IBM Security 2019
////

@import '../../globals/index';

@mixin search-bar {
  display: flex;

  &__scopes.#{$prefix}--list-box {
    flex: 0 0 carbon--mini-units(28);
    border: 0;

    &,
    .#{$prefix}--list-box__field {
      height: 100%;
      max-height: unset;
    }
  }
}
