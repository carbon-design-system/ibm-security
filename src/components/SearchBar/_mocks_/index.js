/**
 * @file Search bar mocks.
 * @copyright IBM Security 2019
 */

const scopes = [
  {
    id: 'scope-2',
    name: 'Scope 2',
  },
  {
    id: 'scope-1',
    name: 'Scope 1',
  },
  {
    id: 'scope-3',
    name: 'Scope 3',
  },
];

export default {
  submitLabel: 'Submit',
  placeHolderText: 'Placeholder',
  labelText: 'Label',
  clearButtonLabelText: 'Clear',
  value: 'Initial value',
  scopes,
  selectedScopes: [scopes[0]],
  scopesTypeLabel: 'Scopes',
  allSelectedScopesLabel: 'All scopes',
};
