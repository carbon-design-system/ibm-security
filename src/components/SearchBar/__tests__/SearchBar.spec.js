/**
 * @file Search bar tests.
 * @copyright IBM Security 2019
 */

import { shallow } from 'enzyme';
import React from 'react';

import { SearchBar } from '../../..';
import props from '../_mocks_';

import { namespace } from '../SearchBar';

const {
  clearButtonLabelText,
  labelText,
  placeHolderText,
  selectedScopes,
  scopes,
  scopesTypeLabel,
  submitLabel,
  value,
} = props;

const regularProps = {
  submitLabel,
  placeHolderText,
  labelText,
  clearButtonLabelText,
};

const { fn } = jest;

describe('SearchBar', () => {
  let searchBar;
  let onChange;

  beforeEach(() => {
    onChange = fn();

    searchBar = shallow(<SearchBar {...regularProps} onChange={onChange} />);
  });

  describe('Rendering', () => {
    it('renders', () => {
      expect(searchBar).toMatchSnapshot();
    });

    it('renders an initial value', () => {
      searchBar.setProps({ value });
      expect(searchBar).toMatchSnapshot();
    });

    it('renders scopes', () => {
      searchBar.setProps({ scopes, scopesTypeLabel });
      expect(searchBar).toMatchSnapshot();
    });

    it('renders selected scopes', () => {
      searchBar.setProps({ scopes, scopesTypeLabel, selectedScopes });

      expect(searchBar).toMatchSnapshot();
    });

    it('disables the submit button with no value', () => {
      searchBar.setProps({ value: null });

      expect(searchBar).toMatchSnapshot();
    });

    it('disables the submit button with no selected scopes', () => {
      searchBar.setProps({
        scopes,
        scopesTypeLabel,
        selectedScopes: [],
      });

      expect(searchBar).toMatchSnapshot();
    });
  });

  describe('Events', () => {
    const simulate = (event, parameters, className = '') =>
      searchBar.find(`.${namespace}${className}`).simulate(event, parameters);

    it('calls the change handler when the value changes', () => {
      simulate('change', { target: { value } }, '__input');

      expect(onChange).toHaveBeenCalledWith({ value });
    });

    it('calls the change handler when the scope changes', () => {
      searchBar.setProps({ scopes, scopesTypeLabel });

      simulate('change', { selectedItems: selectedScopes }, '__scopes');

      expect(onChange).toHaveBeenCalledWith({
        selectedScopes,
        value: SearchBar.defaultProps.value,
      });
    });

    it('calls the submit handler when the submit button is selected', () => {
      const onSubmit = fn();
      searchBar.setProps({ onSubmit, value });

      simulate('submit', { preventDefault: fn() });

      expect(onSubmit).toHaveBeenCalledWith({ value });
    });

    it('prevents the default submit action', () => {
      const preventDefault = fn();

      simulate('submit', { preventDefault });

      expect(preventDefault).toHaveBeenCalledTimes(1);
    });
  });
});
