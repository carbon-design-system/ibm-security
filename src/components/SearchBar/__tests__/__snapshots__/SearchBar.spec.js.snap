// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchBar Rendering disables the submit button with no selected scopes 1`] = `
<form
  className="security--search-bar"
  onSubmit={[Function]}
>
  <MultiSelect
    className="security--search-bar__scopes"
    compareItems={[Function]}
    disabled={false}
    id="security--search-bar__multi-select"
    initialSelectedItems={Array []}
    itemToString={[Function]}
    items={
      Array [
        Object {
          "id": 0,
          "name": "Scope 1",
        },
        Object {
          "id": 1,
          "name": "Scope 2",
        },
        Object {
          "id": 2,
          "name": "Scope 3",
        },
      ]
    }
    label="Scopes"
    light={false}
    locale="en"
    name="search-scopes"
    onChange={[Function]}
    open={false}
    selectionFeedback="top-after-reopen"
    sortItems={[Function]}
    title={false}
    type="default"
  />
  <Search
    className="security--search-bar__input"
    closeButtonLabelText="Clear"
    labelText="Label"
    name="search-input"
    onChange={[Function]}
    placeHolderText="Placeholder"
    showIcon={false}
    type="text"
    value=""
  />
  <Button
    className="security--search-bar__submit"
    disabled={true}
    kind="primary"
    loading={false}
    name="search-submit"
    tabIndex={0}
    type="submit"
  >
    Submit
  </Button>
</form>
`;

exports[`SearchBar Rendering disables the submit button with no value 1`] = `
<form
  className="security--search-bar"
  onSubmit={[Function]}
>
  <Search
    className="security--search-bar__input"
    closeButtonLabelText="Clear"
    labelText="Label"
    name="search-input"
    onChange={[Function]}
    placeHolderText="Placeholder"
    showIcon={false}
    type="text"
    value={null}
  />
  <Button
    className="security--search-bar__submit"
    disabled={true}
    kind="primary"
    loading={false}
    name="search-submit"
    tabIndex={0}
    type="submit"
  >
    Submit
  </Button>
</form>
`;

exports[`SearchBar Rendering renders 1`] = `
<form
  className="security--search-bar"
  onSubmit={[Function]}
>
  <Search
    className="security--search-bar__input"
    closeButtonLabelText="Clear"
    labelText="Label"
    name="search-input"
    onChange={[Function]}
    placeHolderText="Placeholder"
    showIcon={false}
    type="text"
    value=""
  />
  <Button
    className="security--search-bar__submit"
    disabled={true}
    kind="primary"
    loading={false}
    name="search-submit"
    tabIndex={0}
    type="submit"
  >
    Submit
  </Button>
</form>
`;

exports[`SearchBar Rendering renders an initial value 1`] = `
<form
  className="security--search-bar"
  onSubmit={[Function]}
>
  <Search
    className="security--search-bar__input"
    closeButtonLabelText="Clear"
    labelText="Label"
    name="search-input"
    onChange={[Function]}
    placeHolderText="Placeholder"
    showIcon={false}
    type="text"
    value="Initial value"
  />
  <Button
    className="security--search-bar__submit"
    disabled={false}
    kind="primary"
    loading={false}
    name="search-submit"
    tabIndex={0}
    type="submit"
  >
    Submit
  </Button>
</form>
`;

exports[`SearchBar Rendering renders scopes 1`] = `
<form
  className="security--search-bar"
  onSubmit={[Function]}
>
  <MultiSelect
    className="security--search-bar__scopes"
    compareItems={[Function]}
    disabled={false}
    id="security--search-bar__multi-select"
    initialSelectedItems={Array []}
    itemToString={[Function]}
    items={
      Array [
        Object {
          "id": 0,
          "name": "Scope 1",
        },
        Object {
          "id": 1,
          "name": "Scope 2",
        },
        Object {
          "id": 2,
          "name": "Scope 3",
        },
      ]
    }
    label="Scopes"
    light={false}
    locale="en"
    name="search-scopes"
    onChange={[Function]}
    open={false}
    selectionFeedback="top-after-reopen"
    sortItems={[Function]}
    title={false}
    type="default"
  />
  <Search
    className="security--search-bar__input"
    closeButtonLabelText="Clear"
    labelText="Label"
    name="search-input"
    onChange={[Function]}
    placeHolderText="Placeholder"
    showIcon={false}
    type="text"
    value=""
  />
  <Button
    className="security--search-bar__submit"
    disabled={true}
    kind="primary"
    loading={false}
    name="search-submit"
    tabIndex={0}
    type="submit"
  >
    Submit
  </Button>
</form>
`;

exports[`SearchBar Rendering renders selected scopes 1`] = `
<form
  className="security--search-bar"
  onSubmit={[Function]}
>
  <MultiSelect
    className="security--search-bar__scopes"
    compareItems={[Function]}
    disabled={false}
    id="security--search-bar__multi-select"
    initialSelectedItems={
      Array [
        Object {
          "id": 0,
          "name": "Scope 1",
        },
      ]
    }
    itemToString={[Function]}
    items={
      Array [
        Object {
          "id": 0,
          "name": "Scope 1",
        },
        Object {
          "id": 1,
          "name": "Scope 2",
        },
        Object {
          "id": 2,
          "name": "Scope 3",
        },
      ]
    }
    label="Scopes"
    light={false}
    locale="en"
    name="search-scopes"
    onChange={[Function]}
    open={false}
    selectionFeedback="top-after-reopen"
    sortItems={[Function]}
    title={false}
    type="default"
  />
  <Search
    className="security--search-bar__input"
    closeButtonLabelText="Clear"
    labelText="Label"
    name="search-input"
    onChange={[Function]}
    placeHolderText="Placeholder"
    showIcon={false}
    type="text"
    value=""
  />
  <Button
    className="security--search-bar__submit"
    disabled={true}
    kind="primary"
    loading={false}
    name="search-submit"
    tabIndex={0}
    type="submit"
  >
    Submit
  </Button>
</form>
`;
